app/src/androidTest/java/org/kiwix/kiwixmobile/testutils/TestUtils.kt

```diff
@@ -0,0 +1,138 @@
+import androidx.test.espresso.matcher.BoundedMatcher
+import androidx.test.platform.app.InstrumentationRegistry
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.kiwix.kiwixmobile.core.entity.LibraryNetworkEntity
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.IOException
+import java.text.SimpleDateFormat
+import java.util.Date
+
+/**
+ * Created by mhutti1 on 07/04/17.
+ */
+object TestUtils {
+  private const val TAG = "TESTUTILS"
+  @JvmField var TEST_PAUSE_MS = 250
+  var TEST_PAUSE_MS_FOR_SEARCH_TEST = 1000
+
+  /*
+    TEST_PAUSE_MS is used as such:
+        BaristaSleepInteractions.sleep(TEST_PAUSE_MS);
+    The number 250 is fairly arbitrary. I found 100 to be insufficient,
+        and 250 seems to work on all devices I've tried.
+
+    The sleep combats an intermittent issue caused by
+        tests executing before the app/activity is ready.
+    This isn't necessary on all devices (particularly more recent ones),
+        however I'm unsure if
+    it's speed related, or Android Version related.
+   */
+  private fun hasStoragePermission(): Boolean {
+    return ContextCompat.checkSelfPermission(
+      InstrumentationRegistry.getInstrumentation().targetContext,
+      Manifest.permission.READ_EXTERNAL_STORAGE
+    ) == PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(
+      InstrumentationRegistry.getInstrumentation().targetContext,
+      Manifest.permission.WRITE_EXTERNAL_STORAGE
+    ) == PackageManager.PERMISSION_GRANTED
+  }
+
+  @JvmStatic fun allowPermissionsIfNeeded() {
+    if (Build.VERSION.SDK_INT >= 23 && !hasStoragePermission()) {
+      val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+      val allowPermissions =
+        device.findObject(
+          UiSelector().clickable(true)
+            .checkable(false).index(1)
+        )
+      if (allowPermissions.exists()) {
+        try {
+          allowPermissions.click()
+        } catch (e: UiObjectNotFoundException) {
+        }
+      }
+    }
+  }
+
+  @JvmStatic fun captureAndSaveScreenshot(name: String?) {
+    val screenshotDir = File(
+      Environment.getExternalStorageDirectory().toString() +
+        "/Android/data/KIWIXTEST/Screenshots"
+    )
+    if (!screenshotDir.exists()) {
+      if (!screenshotDir.mkdirs()) {
+        return
+      }
+    }
+    val timestamp = SimpleDateFormat("ddMMyyyy_HHmm").format(Date())
+    val fileName = "TEST_${timestamp}_$name.png"
+    val outFile = File(screenshotDir.path + File.separator + fileName)
+    val screenshot = Screenshot.capture().bitmap ?: return
+    try {
+      val fos = FileOutputStream(outFile)
+      screenshot.compress(Bitmap.CompressFormat.PNG, 90, fos)
+      fos.close()
+    } catch (e: FileNotFoundException) {
+      Log.w(TAG, "Failed to save Screenshot", e)
+    } catch (e: IOException) {
+      Log.w(TAG, "Failed to save Screenshot", e)
```

Comment:
Perhaps a more meaningful message in each exception

Reply:
hi @gouri-panda 
done